### PR TITLE
fix(react): iconButtons have reiterate classNames

### LINKED ISSUE #15459: [Bug]: IconButtons have reiterate classNames in a button HTML element.
### Package

@carbon/react

### Browser

Chrome

### Package version

^1.46.1

### React version

_No response_

### Description

Icon buttons have duplicate classNames in their button HTML element. (the picture below)
In this example, this IconButton's className is `"cds--btn--icon-only cds--btn cds--btn--primary cds--btn-
-icon-only cds--btn cds--btn--primary"`.
As you can see, we iterating className twice.

249D1C93-9E44-4C2E-9630-F0555CF90CC4_1_201_a

### Reproduction/example

https://react.carbondesignsystem.com/?path=/story/components-button--icon-button

### Steps to reproduce

Go to the storybook, we will see this example that I attached in the description.
https://react.carbondesignsystem.com/?path=/story/components-button--icon-button

### Suggested Severity

Severity 4 = Unrelated to a user task, has a workaround or does not need a workaround.

### Application/PAL

_No response_

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/carbon-design-system/carbon/blob/f555616971a03fd454c0f4daea184adf41fff05b/.github/CODE_OF_CONDUCT.md)
- [X] I checked the [current issues](https://github.com/carbon-design-system/carbon/issues) for duplicate problems

### PR DESCRIPTION
**Pull request by harukikuri**
Closes #15459

Icon buttons have duplicate classNames in their button HTML element because of the circular dependency.

#### Changelog

**New**
- ButtonBase.tsx

**Changed**
- Button.tsx
- index.tsx (IconButton folder)

#### Testing / Reviewing

To avoid the circular dependency, introduced a new ButtonBase.

[OLD]
Button > IconButton > Button    <--- This is the cause of "circular dependency".
[NEW]
Button > IconButton > ButtonBase

In IconButton, we used to import the Button component once again since we needed to rely on the logic(i.e. generating some classNames, ...etc.) that resides in Button.tsx.
However, this logic in Button.tsx is seperatable. Upon investigation, I noticed that the logic was a mixture of the logic for the Button component(not `hasIconOnly`) and the IconButton component. So, by splitting this logic up and creating a new ButtonBase component to migrate the Button component's logic from Button.tsx.

Here is a brief explanation for this.
Button.tsx … If-else condition for deciding whether the Button component or IconButton component. 
ButtonBase.tsx … The Button component's logic exists
IconButton.tsx … Export the IconButton component

So in a nutshell, Extracted a logic for the Button component into ButtonBase.tsx from Button.tsx to leave Button.tsx file's responsibility only deciding whether the Button component or IconButton component.

As a pic below, we successfully removed duplicated classNames from the Button component. Also, the circular dependency during the build as well!(#14399)
Screenshot 2024-01-28 at 1 23 00 PM (2)

### CODE REVIEW

**Comment by kuri-sun** [comment]
Sorry I have to update this PR, I will let you know if it's ready! Thank you!

**Comment by kuri-sun** [comment]
I updated the PR, so it's ready! Thank you for your review!

**Review by guidari — APPROVED** [review]
LGTM! I like the solution to fix the circular dependency issue.  🚀

Let's wait for another dev to review it!

**Comment by kuri-sun** [comment]
Thanks for the review, @guidari! 😄

**Review by tay1orjones — CHANGES_REQUESTED** [review]
Thanks for taking a look at this! This change will impact Storybook's ability to autogenerate the component API tables. 

Right now the defaults for Button are not coming through: https://deploy-preview-15626--v11-carbon-react.netlify.app/?path=/docs/components-button--overview#component-api

image

We won't want to add BaseButton to the table because it's not exported. I'm not sure how best to fix this without specifying the default for every prop in the storybook args/argTypes.

**Comment by kuri-sun** [comment]
Sorry everyone! I had a long break so I didn't have time to look at this one!
Sure, @tay1orjones!
I will renew this with taking Storybook into consideration!
Thank you for your patience! 😄

**Comment by kuri-sun** [comment]
> Thanks for taking a look at this! This change will impact Storybook's ability to autogenerate the component API tables.
> 
> Right now the defaults for Button are not coming through: https://deploy-preview-15626--v11-carbon-react.netlify.app/?path=/docs/components-button--overview#component-api
> 
> image
> 
> We won't want to add BaseButton to the table because it's not exported. I'm not sure how best to fix this without specifying the default for every prop in the storybook args/argTypes.

Hi @tay1orjones!, 
I just checked the component table and as far as I saw, I don't think any changes are detected between the current Button component's props table and this branch's Button component's props table.

Here I attached the preview of this branch's Button components' props table. Please take a look. Thanks!

Screenshot 2024-03-18 at 11 12 08 AM

Screenshot 2024-03-18 at 11 12 30 AM

Screenshot 2024-03-18 at 11 12 44 AM

Screenshot 2024-03-18 at 11 12 00 AM

**Review by alisonjoseph — APPROVED** [review]
This LGTM, @tay1orjones are you still seeing an issue with prop table?

**Review by tay1orjones — APPROVED** [review]
Didn't realize the defaults were broken in prod too. We can fix them in a different issue/PR. Thanks!
